Feat : Admin 관련 API 개발

Young-Yeong opened the pull request:
Related to : #82, #81, #83, #84, #85, #88

Young-Yeong committed:
Feat : Admin Entity 관련 API 개발

- AdminController
    - GET /api/v1/admin/login
        - request : AdminLoginDto adminLoginDto
        - response : Boolean

- AdminLoginDto
    - 관리자 로그인 Dto

- AdminService
    - loginAdmin
        - 입력받은 adminLoginDto을 통해 로그인 여부 반환

Related to : #82

Young-Yeong committed:
Feat : Notice Entity 관련 API 개발

- Notice
    - updateNotice
        - noticeDto에 맞게 update

- NoticeController
    - postNotice
        - POST /api/v1/branches/notices
            - request : NoticeDto noticedto
                        String username
            - response : Long

    - patchNotice
        - PATCH /api/v1/branches/notices
            - request : NoticeDto noticedto
                        String username
                        Long noticeId
            - response : Boolean

    - deleteNotice
        - DELETE /api/v1/branches/notices
            - request : String username
                        Long noticeId
            - response : Boolean

- NoticeDto
    - 공지사항 Dto

- NoticeService
    - postNotice
        - 입력받은 noticeDto 저장
    - patchNotice
        - 입력받은 noticeId로 검색 후 같은 지점 관리자인지 확인 후 noticeDto에 맞게 변경
    - deleteNotice
        - 입력받은 noticeId로 검색 후 같은 지점 관리자인지 확인 후 삭제

- Branch
    - EqualsAndHashCode 어노테이션 추가

Related to : #81

Young-Yeong committed:
Feat : CarSpecification Entity 관련 API 개발

- CarSpecController
    - postCarSpec
        - POST /api/v1/car-spec
            - request : CarSpecRequestDto carSpecRequestDto
                        String username
            - response : Long

    - deleteCarSpec
        - DELETE /api/v1/car-spec
            - request : Long carSpecId
                        String username
            - response : Boolean

- CarSpecification
    - EqualsAndHashCode 어노테이션 추가

- CarSpecificationRepository
    - findByCarBrandAndCarNameAndOilTypeAndReleaseDateAndTransmission
        - CarBrand, CarName, OilType, ReleaseDate, Transmission이 모두 같은 레코드 반환

- CarSpecRequestDto
    - 차량 스펙 요청 Dto

- CarSpecService
    - postCarSpec
        - 이미 등록된 차량 스펙인지 확인 후 등록

    - deleteCarSpec
        - 입력받은 carSpecId에 해당하는 차량 스펙 삭제

Related to : #83

Young-Yeong committed:
Feat : Point Entity 관련 API 개발

- PointController
    - getPoint
        - GET /api/v1/points/users
            - request : String adminUsername
                        String username
            - response : Integer

    - postPoint
        - POST /api/v1/points/users
            - request : String adminUsername
                        PointRequestDto pointRequestDto
            - response : Long

- PointRequestDto
    - 포인트 요청 Dto

- PointService
    - getPoint
        - 검증 후 유저가 가진 포인트 반환

    - postPoint
        - 검증 후 PointType에 맞게 데이터 추가

Related to : #84

Young-Yeong committed:
Feat : Key Entity 관련 API 개발

- Key
    - ToString 어노테이션 추가
    - Builder 추가
    - updateKey 메소드 추가

- KeyController
    - postKey
        - POST /api/v1/branches/cars/keys
            - request : KeyDto keyDto
                        String adminUsername
            - response : Long
    - patchKey
        - PATCH /api/v1/branches/cars/keys
            - request : KeyDto keyDto
                        String adminUsername
                        Long keyId
            - response : Boolean
    - deleteKey
        - DELETE /api/v1/branches/cars/keys
            - request : String adminUsername
                        Long keyId
            - response : Boolean
- KeyDto
    - 차 키 관련 Dto

- KeyRepository
    - JPA 상속

- KeyService
    - postKey
        - 검증 후 차 키 저장
    - patchKey
        - 검증 후 keyId로 검색 후 keyDto에 맞게 수정
    - deleteKey
        - 검증 후 keyId로 검색 후 삭제

- KeyStorageRepository
    - JPA 상속

- KioskRepository
    - JPA 상속

Related to : #85

Young-Yeong committed:
Feat : Cors WildCard 설정

- WebMvcConfig
    - Cors wildcard 설정
Related to : #88

Capoomaru commented on src/main/java/com/yu/yurentcar/domain/branch/entity/Notice.java:
```diff
@@ -62,4 +63,17 @@ public Notice(Long noticeId, @NotNull Branch branch, @NotNull Admin admin, @NotN
         this.finishDate = finishDate;
     }
+
+    public Notice updateNotice(NoticeDto noticeDto, Admin admin){
+        this.title = noticeDto.getTitle();
+        if(startDate != null)
+            this.startDate = noticeDto.getStartDate();
+        if(finishDate != null)
+            this.finishDate = noticeDto.getFinishDate();
+        this.description = noticeDto.getDescription();
+        if(photoUrl != null)
+            this.photoUrl = noticeDto.getPhotoUrl();
+        this.admin = admin;
+        return this;
```
Dynamic update good!

Capoomaru commented on src/main/java/com/yu/yurentcar/domain/branch/service/NoticeService.java:
```diff
@@ -43,4 +49,48 @@ public NoticeContentResponseDto getNoticeContentByNoticeId(Long noticeId) {
+        Optional<Admin> lookupAdmin = adminRepository.findByUsername(username);
+        Optional<Notice> lookupNotice = noticeRepository.findById(noticeId);
+        if (lookupAdmin.isEmpty()) throw new RuntimeException("없는 관리자입니다.");
+        //수정하려는 관리자와 작성한 관리자가 같은 지점 관리자인지 확인
+        if (!lookupAdmin.get().getBranch().equals(lookupNotice.get().getBranch()))
+            throw new RuntimeException("다른 지점 관리자입니다. 권한이 없습니다.");
```
나중에 시큐리티를 엄청 세밀하게 조정하면 빠져도 될 것 같아요

Capoomaru commented on src/main/java/com/yu/yurentcar/domain/branch/service/NoticeService.java:
```diff
@@ -43,4 +49,48 @@ public NoticeContentResponseDto getNoticeContentByNoticeId(Long noticeId) {
+        //수정하려는 관리자와 작성한 관리자가 같은 지점 관리자인지 확인
+        if (!lookupAdmin.get().getBranch().equals(lookupNotice.get().getBranch()))
+            throw new RuntimeException("다른 지점 관리자입니다. 권한이 없습니다.");
+        noticeRepository.save(lookupNotice.orElseThrow(() -> new RuntimeException("없는 공지사항입니다."))
+                .updateNotice(noticeDto, lookupAdmin.get()));
```
throw가 발생할 수 있으니 가독성 등을 고려해서 분리하는게 나을 것 같아요
```java
//before
noticeRepository.save(lookupNotice.orElseThrow(() -> new RuntimeException("없는 공지사항입니다."))
        .updateNotice(noticeDto, lookupAdmin.get()));
//after
Notice notice = lookupNotice.orElseThrow(() -> new RuntimeException("없는 공지사항입니다.")
          .updateNotice(noticeDto, lookupAdmin.get());
noticeRepository.save(notice);
```

Capoomaru commented on src/main/java/com/yu/yurentcar/domain/branch/service/NoticeService.java:
```diff
@@ -43,4 +49,48 @@ public NoticeContentResponseDto getNoticeContentByNoticeId(Long noticeId) {
+    public Boolean deleteNotice(String username, Long noticeId) {
+        Optional<Admin> lookupAdmin = adminRepository.findByUsername(username);
+        Optional<Notice> lookupNotice = noticeRepository.findById(noticeId);
+        if (lookupAdmin.isEmpty()) throw new RuntimeException("없는 관리자입니다.");
+        if (lookupNotice.isEmpty()) throw new RuntimeException("없는 공지사항입니다.");
+        //수정하려는 관리자와 작성한 관리자가 같은 지점 관리자인지 확인
+        if (!lookupAdmin.get().getBranch().equals(lookupNotice.get().getBranch()))
+            throw new RuntimeException("다른 지점 관리자입니다. 권한이 없습니다.");
```
예외사항 체킹 좋아요

Capoomaru commented on src/main/java/com/yu/yurentcar/domain/car/controller/CarSpecController.java:
```diff
@@ -19,17 +18,29 @@ public class CarSpecController {
+    @PostMapping
+    public ResponseEntity<Long> postCarSpec(@RequestBody CarSpecRequestDto carSpecRequestDto, @RequestParam String username) {
+        log.info("postCarSpec = " + carSpecRequestDto);
+        return ResponseEntity.status(HttpStatus.OK).body(carSpecService.postCarSpec(carSpecRequestDto, username));
```
가급적 body 내용을 변수로 만들고 넣는게 좋을 것 같아요 어떤 정보가 들어가는지 알기 어려워요

Capoomaru commented on src/main/java/com/yu/yurentcar/domain/car/entity/CarSpecification.java:
```diff
@@ -19,6 +16,7 @@
 @NoArgsConstructor
 @Table(name = "car_specification")
 @ToString
+@EqualsAndHashCode
```
비교처리 좋아요 👍

Capoomaru commented on src/main/java/com/yu/yurentcar/domain/car/entity/Key.java:
```diff
@@ -35,4 +34,20 @@ public class Key extends BaseTimeEntity {
+        this.state = keyState;
+    }
+
+    public Key updateKey(Car car, KeyStorage keyStorage, String rfid, KeyState keyState) {
+        this.car = car;
+        this.keyStorage = keyStorage;
+        this.rfid = rfid;
+        this.state = keyState;
+        return this;
+    }
```
업데이트할 때 무조건 `car`, `keyStore`, `rfid`, `keyState` 가 모두 변경되서 들어오는 건가요?

그게 아니라면, null check가 필요해보여요.

Capoomaru commented on src/main/java/com/yu/yurentcar/domain/car/service/CarSpecService.java:
```diff
@@ -34,4 +43,44 @@ public List<String> getOilTypes() {
+
+        Optional<CarSpecification> lookupCarSpec = carSpecificationRepository
+                .findByCarBrandAndCarNameAndOilTypeAndReleaseDateAndTransmission(
+                        Enum.valueOf(CarBrand.class, carSpecRequestDto.getCarBrand()),
```
CarBrand는 Converter로 변환이 불가능한가요?

Capoomaru commented on src/main/java/com/yu/yurentcar/domain/car/service/KeyService.java:
```diff
@@ -0,0 +1,82 @@
+
+    @Transactional
+    public Boolean deleteKey(String adminUsername, Long keyId) {
+        Optional<Admin> lookupAdmin = adminRepository.findByUsername(adminUsername);
+        if (lookupAdmin.isEmpty()) throw new RuntimeException("없는 관리자입니다.");
```
위에서 쓰던 `orElseThrow`을 이용해도 될 것 같아요.
```java
//before
Optional<Admin> lookupAdmin = adminRepository.findByUsername(adminUsername);
        if (lookupAdmin.isEmpty()) throw new RuntimeException("없는 관리자입니다.");
//after
Admin lookupAdmin  = adminRepository.findByUsername(adminUsername)
        .orElseThrow(() => throw new RuntimeException("없는 관리자입니다."));
```

Capoomaru commented on src/main/java/com/yu/yurentcar/security/config/WebMvcConfig.java:
```diff
@@ -0,0 +1,19 @@
+@EnableWebMvc
+public class WebMvcConfig implements WebMvcConfigurer {
+
+    @Override
+    public void addCorsMappings(CorsRegistry registry) {
+        registry.addMapping("/**")
+                .allowedOrigins("http://localhost:3000")
+                .allowedMethods("OPTIONS","GET","POST","PUT","DELETE");
+    }
```
Cors 터졌나요...? 다른 방식으로 해결했어야 할 것 같아요 :(

Capoomaru reviewed (commented):
혼자서 너무 고생하셧습니다 👏

Capoomaru commented on src/main/java/com/yu/yurentcar/domain/branch/service/NoticeService.java:
```diff
@@ -25,14 +25,14 @@ public class NoticeService {
     @Transactional(readOnly = true)
-    public List<NoticeResponseDto> getNoticesByBranchName(String province, String branchName) {
+    public List<NoticeResponseDto> getNoticesByBranchName(String province, String branchName, Integer count) {
         SiDoType siDo;
         try {
             siDo = EnumValueConvertUtils.ofDesc(SiDoType.class, province);
         } catch (RuntimeException e) {
             siDo = null;
         }
-        return noticeRepository.getNoticesByBranchName(siDo, branchName);
+        return noticeRepository.getNoticesByBranchName(siDo, branchName, count);
```
count는 조회수인가요?

Young-Yeong committed:
Fix : #89 수정

- CarBrand
    - EnumValue 상속

- CarSpecService
    - CarBrand Converter 적용

- Key
    - null 체크 적용

- NoticeService
    - 가독성 개선

Related to : #89

Capoomaru commented on src/main/java/com/yu/yurentcar/domain/branch/service/NoticeService.java:
```diff
@@ -73,11 +73,11 @@ public Boolean patchNotice(NoticeDto noticeDto, String username, Long noticeId)
             throw new RuntimeException("다른 지점 관리자입니다. 권한이 없습니다.");
-        noticeRepository.save(lookupNotice.orElseThrow(() -> new RuntimeException("없는 공지사항입니다."))
-                .updateNotice(noticeDto, lookupAdmin.get()));
+        noticeRepository.save(lookupNotice.get().updateNotice(noticeDto, lookupAdmin.get()));
```
```java
Notice notice = lookupNotice.get().updateNotice(noticeDto, lookupAdmin.get());
noticeRepository.save(notice);
```
이렇게 수정하라는 의미였어요. 그래도 가독성이 훨씬 좋아보여요 👍

Capoomaru reviewed (approved):
대부분 잘 반영된 것 같네요!
LGTM
